aquarius/events/processors.py

```diff
@@ -0,0 +1,248 @@
+#
+# Copyright 2021 Ocean Protocol Foundation
+# SPDX-License-Identifier: Apache-2.0
+#
+from abc import ABC
+from datetime import datetime
+from eth_utils import add_0x_prefix, remove_0x_prefix
+import json
+import logging
+
+from plecos.plecos import is_valid_dict_remote, list_errors_dict_remote
+from aquarius.app.auth_util import compare_eth_addresses
+from aquarius.app.util import (
+    DATETIME_FORMAT,
+    format_timestamp,
+    get_metadata_from_services,
+    init_new_ddo,
+    list_errors,
+    validate_data,
+)
+from aquarius.events.constants import EVENT_METADATA_CREATED
+from aquarius.events.util import get_datatoken_info
+from aquarius.events.decryptor import Decryptor
+
+logger = logging.getLogger(__name__)
+debug_log = logger.debug
+
+
+class EventProcessor(ABC):
+    def __init__(self, event, oceandb, web3, ecies_account, allowed_publishers):
+        self.event = event
+        self.did = f"did:op:{remove_0x_prefix(self.event.args.dataToken)}"
+        self.block = event.blockNumber
+        self.txid = self.event.transactionHash.hex()
+        self.contract_address = self.event.address
+        self.sender_address = self.event.args.get(
+            "createdBy", self.event.args.get("updatedBy")
+        )
+        self.flags = event.args.get("flags", None)
+        self.rawddo = event.args.get("data", None)
+
+        self._oceandb = oceandb
+        self._web3 = web3
+        self.decryptor = Decryptor(ecies_account)
+        self.allowed_publishers = allowed_publishers
+
+        blockInfo = self._web3.eth.getBlock(self.event.blockNumber)
+        self.timestamp = blockInfo["timestamp"]
+
+
+class MetadataCreatedProcessor(EventProcessor):
+    def is_publisher_allowed(self, publisher_address):
+        logger.debug(f"checking allowed publishers: {publisher_address}")
+        if not self.allowed_publishers:
+            return True
+
+        publisher_address = self._web3.toChecksumAddress(publisher_address)
+        return publisher_address in self.allowed_publishers
+
+    def make_record(self, data):
+        _record = init_new_ddo(data, self.timestamp)
+        # this will be used when updating the doo
```

Comment:
Typo. I believe doo should be ddo. Also, what "this" is this comment referring to?